pkg/skaffold/deploy/label.go

```diff
@@ -0,0 +1,152 @@
+/*
+Copyright 2018 The Skaffold Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package deploy
+
+import (
+	"time"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/constants"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/kubernetes"
+	"github.com/sirupsen/logrus"
+
+	clientgo "k8s.io/client-go/kubernetes"
+
+	appsv1 "k8s.io/api/apps/v1"
+	appsv1beta1 "k8s.io/api/apps/v1beta1"
+	appsv1beta2 "k8s.io/api/apps/v1beta2"
+	corev1 "k8s.io/api/core/v1"
+	extensionsv1beta1 "k8s.io/api/extensions/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// retry 3 times to give the object time to propagate to the API server
+const tries int = 3
+const sleeptime time.Duration = 300 * time.Millisecond
+
+func LabelDeployResults(labels map[string]string, results []Artifact) {
+	// use the kubectl client to update all k8s objects with a skaffold watermark
+	client, err := kubernetes.Client()
+	if err != nil {
+		logrus.Warnf("error retrieving kubernetes client: %s", err.Error())
+		return
+	}
+	for _, res := range results {
+		err = nil
+		for i := 0; i < tries; i++ {
+			if err = updateRuntimeObject(client, labels, res); err == nil {
+				break
+			}
+			time.Sleep(sleeptime)
+		}
+		if err != nil {
+			logrus.Warnf("error adding label to runtime object: %s", err.Error())
+		}
+	}
+}
+
+func addSkaffoldLabels(labels map[string]string, m *metav1.ObjectMeta) {
+	if m.Labels == nil {
+		m.Labels = map[string]string{}
+	}
+	for k, v := range labels {
+		m.Labels[k] = v
+	}
+}
+
+func retrieveNamespace(ns string, m metav1.ObjectMeta) string {
+	if ns != "" {
+		return ns
+	}
+	if m.Namespace != "" {
+		return m.Namespace
+	}
+	return "default"
+}
+
+func updateRuntimeObject(client clientgo.Interface, labels map[string]string, res Artifact) error {
```

Comment:
I believe the dynamic client should make this easy. 

https://github.com/kubernetes/client-go/blob/master/dynamic/interface.go

Reply:
Looks like client-go itself is discouraging use of the dynamic client: https://github.com/kubernetes/client-go/issues/394. Even though it looks a little messy I think sticking to the typed APIs is safer and probably a better idea

Reply:
I think we should use it anyways. Lets keep it out of scope for this PR though.